deeplearning4j-dataimport-solrj/src/test/java/org/deeplearning4j/nn/dataimport/solr/client/solrj/io/stream/TupleStreamDataSetIteratorTest.java

```diff
@@ -0,0 +1,181 @@
+              "luminance_f", Float.toString(luminance));
+
+            updateRequest.add(doc);
+            ++numDocs;
+
+          }
+        }
+      }
+    }
+
+    // make the update request
+    updateRequest.commit(cluster.getSolrClient(), "mySolrCollection");
+  }
+
+  private static class CountingIterationListener extends ScoreIterationListener {
+
+    private int numIterationsDone = 0;
+
+    public CountingIterationListener() {
+      super(1);
+    }
+
+    public int numIterationsDone() {
+      return numIterationsDone;
+    }
+
+    @Override
+    public void iterationDone(Model model, int iteration, int epoch) {
+      super.iterationDone(model, iteration, epoch);
+      ++numIterationsDone;
+    }
+
+  }
+
+  @Test
+  public void iterateTest() throws Exception {
+
+      try (final TupleStreamDataSetIterator
+      tsdsi = new TupleStreamDataSetIterator(
+        123 /* batch */,
+        new String[] { "pie" },
+        new String[] { "answer" },
+        "tuple(greeting=\"hello world\",pie=3.14,answer=42)",
+        null)) {
+
+      assertTrue(tsdsi.hasNext());
+      final DataSet ds = tsdsi.next();
+
+      assertEquals(1, ds.getFeatures().length());
+      assertEquals(3.14f, ds.getFeatures().getFloat(0), 0.0f);
+
+      assertEquals(1, ds.getLabels().length());
+      assertEquals(42f, ds.getLabels().getFloat(0), 0.0f);
+
+      assertFalse(tsdsi.hasNext());
+    }
+  }
+
+  @Test
+  public void modelFitTest() throws Exception {
+
+    final MultiLayerNetwork model = new MultiLayerNetwork(
+      new NeuralNetConfiguration.Builder()
+        .list(
+          new OutputLayer.Builder(LossFunction.MSE)
+            .nIn(3)
+            .nOut(1)
+            .weightInit(WeightInit.ONES)
+            .activation(Activation.IDENTITY)
+            .build()
+          )
+        .pretrain(false)
+        .backprop(true)
+        .build()
+      );
+    model.init();
+
+    int batch = 1;
+    for (int ii=1; ii<=5; ++ii) {
+      final CountingIterationListener listener = new CountingIterationListener();
+      model.setListeners(listener);
+      batch *= 2;
+
+      try (final TupleStreamDataSetIterator tsdsi =
+          new TupleStreamDataSetIterator(
+            batch,
+            new String[] { "channel_b_f", "channel_g_f", "channel_r_f" },
+            new String[] { "luminance_f" },
+            "search(mySolrCollection," +
+            "q=\"id:*\"," +
+            "fl=\"id,channel_b_f,channel_g_f,channel_r_f,luminance_f\"," +
+            "sort=\"id asc\"," +
+            "qt=\"/export\")",
+            cluster.getZkClient().getZkServerAddress())) {
+
+        model.fit(tsdsi);
+      }
+
+      assertEquals("numIterationsDone="+listener.numIterationsDone()+" numDocs="+numDocs+" batch="+batch,
+                   (numDocs+(batch-1))/batch, listener.numIterationsDone());
+    }
+
+    Nd4j.getWorkspaceManager().destroyAllWorkspacesForCurrentThread(); // TODO: seems to have no effect?
```

Comment:
Here's some further info re: this _"seems to have no effect?"_ comment, the tests pass but thread leaks are reported:

```
Tests run: 4, Failures: 0, Errors: 2, Skipped: 0, Time elapsed: 98.979 sec <<< FAILURE! - in org.deeplearning4j.nn.dataimport.solr.client.solrj.io.stream.TupleStreamDataSetIteratorTest
org.deeplearning4j.nn.dataimport.solr.client.solrj.io.stream.TupleStreamDataSetIteratorTest  Time elapsed: 25.694 sec  <<< ERROR!
com.carrotsearch.randomizedtesting.ThreadLeakError: 
3 threads leaked from SUITE scope at org.deeplearning4j.nn.dataimport.solr.client.solrj.io.stream.TupleStreamDataSetIteratorTest: 
   1) Thread[id=171, name=JavaCPP Deallocator, state=WAITING, group=TGRP-TupleStreamDataSetIteratorTest]
        at java.lang.Object.wait(Native Method)
        at java.lang.ref.ReferenceQueue.remove(ReferenceQueue.java:143)
        at java.lang.ref.ReferenceQueue.remove(ReferenceQueue.java:164)
        at org.bytedeco.javacpp.Pointer$DeallocatorThread.run(Pointer.java:299)
   2) Thread[id=170, name=Workspace deallocator thread, state=WAITING, group=TGRP-TupleStreamDataSetIteratorTest]
        at java.lang.Object.wait(Native Method)
        at java.lang.ref.ReferenceQueue.remove(ReferenceQueue.java:143)
        at java.lang.ref.ReferenceQueue.remove(ReferenceQueue.java:164)
        at org.nd4j.linalg.memory.provider.BasicWorkspaceManager$WorkspaceDeallocatorThread.run(BasicWorkspaceManager.java:254)
   3) Thread[id=172, name=NativeRandomDeallocator thread 0, state=WAITING, group=TGRP-TupleStreamDataSetIteratorTest]
        at java.lang.Object.wait(Native Method)
        at java.lang.ref.ReferenceQueue.remove(ReferenceQueue.java:143)
        at java.lang.ref.ReferenceQueue.remove(ReferenceQueue.java:164)
        at org.nd4j.rng.deallocator.NativeRandomDeallocator$DeallocatorThread.run(NativeRandomDeallocator.java:77)

org.deeplearning4j.nn.dataimport.solr.client.solrj.io.stream.TupleStreamDataSetIteratorTest  Time elapsed: 25.694 sec  <<< ERROR!
com.carrotsearch.randomizedtesting.ThreadLeakError: 
There are still zombie threads that couldn't be terminated:
   1) Thread[id=172, name=NativeRandomDeallocator thread 0, state=WAITING, group=TGRP-TupleStreamDataSetIteratorTest]
        at java.lang.Object.wait(Native Method)
        at java.lang.ref.ReferenceQueue.remove(ReferenceQueue.java:143)
        at java.lang.ref.ReferenceQueue.remove(ReferenceQueue.java:164)
        at org.nd4j.rng.deallocator.NativeRandomDeallocator$DeallocatorThread.run(NativeRandomDeallocator.java:77)


Results :

Tests in error: 
org.deeplearning4j.nn.dataimport.solr.client.solrj.io.stream.TupleStreamDataSetIteratorTest.org.deeplearning4j.nn.dataimport.solr.client.solrj.io.stream.TupleStreamDataSetIteratorTest
  Run 1: TupleStreamDataSetIteratorTest.org.deeplearning4j.nn.dataimport.solr.client.solrj.io.stream.TupleStreamDataSetIteratorTest » ThreadLeak
  Run 2: TupleStreamDataSetIteratorTest.org.deeplearning4j.nn.dataimport.solr.client.solrj.io.stream.TupleStreamDataSetIteratorTest » ThreadLeak


Tests run: 3, Failures: 0, Errors: 1, Skipped: 0
```

Reply:
@raver119